modeling/modeling.py

```diff
@@ -0,0 +1,204 @@
+
+
+class Training:
+    """
+    Args:
+        data:
+        estimator:
+    """
+    
+    def __init__(self, data:Posts, estimator):
+        self.data = data
+        self.estimator = estimator
+        self.model = None
+        self.threshold = None
+        
+    
+    def calculate_best_threshold(self, y_true: pd.Series, y_pred_proba: np.array) -> float:
+        """Calculate the best threshold value for classification.
+
+        Args:
+            y_pred_proba: Predicted probabilities.
+            y_true: Series with true target labels.
+
+        Returns:
+            best_th: The threshold value that maximizes the f1-score.
+        """
+        best_th = 0.0
+        best_f1 = 0.0
+        for th in np.arange(0.05, 0.96, 0.05):
+            y_pred_temp = predict_with_threshold(y_pred_proba, th)
+            f1_temp = f1_score(y_true, y_pred_temp)
+            if f1_temp > best_f1:
+                best_th = th
+                best_f1 = f1_temp
+        return best_th
+
+
+    def save_model_to_drive(self, model):
+        """
+        """
+        # ToDo get model info
+        logger.info("Saving model in the models folder")
+        t = datetime.now().strftime('%Y-%m-%d_%H%M%S')
+        path = f"models/{model_details['name']}_{mlflow_params['label']}_{t}"
+        save_model(sk_model=model, path=path)
+
+
+    def train(self): #, mlflow_logger):
+        """Trains the estimator.
+
+        If model implements predict_proba, calculate the best cut-off threshold.
+        If the estimator is a GridSearch, it stores the best_params_ in mlflow_params and returns the best_model_ as model. Stopwords must be set via `vectorizer__stop_words`.
+
+        Args:
+            estimator:
+            mlflow_logger:
+
+        Returns:
+            model:
+            mlflow_params:
+            threshold:
+        """
+        mlflow_params = {} #ToDo fix mlflow
+
+        X_train, y_train = self.data.get_X_y(split="train")
+        
+        logger.info(f"Fit model")
+        model = self.estimator.fit(X_train, y_train) #ToDo save to mlflow logger
+
+        # select best threshold if model implements predict_proba
+        if callable(getattr(model, "predict_proba", None)):
+            y_train_proba = model.predict_proba(X_train)[:, 1]
+            self.threshold = self.calculate_best_threshold(y_train, y_train_proba)
+            mlflow_params["threshold"] = self.threshold #ToDo save to mlflow logger
+
+        # store best parameters if model is GridSearch
+        if isinstance(model, GridSearchCV):
+            best_params = model.best_params_
+            if 'vectorizer__stop_words' in best_params.keys() and best_params['vectorizer__stop_words']!=None:
+                best_params['vectorizer__stop_words'] = "NLTK-German"
+            mlflow_params["best_params"] = best_params #ToDo save to mlflow logger
+
+        self.model = model
+        #self.save_model_to_drive(model)
+        return model
```

Comment:
With 
```
model = self.estimator.fit(X_train, y_train)
```
`model` is just the same as `self.estimator`
I guess you want something like
```
model_best = self.estimator.best_estimator_
``` 
after the fitting?

Relevant text from the [sklearn documentation](https://scikit-learn.org/stable/developers/develop.html#fitting) 

> The method should return the object (self). This pattern is useful to be able to implement quick one liners in an IPython session such as:

Reply:
done